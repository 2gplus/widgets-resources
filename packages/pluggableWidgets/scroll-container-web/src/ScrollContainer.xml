<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.scrollcontainer.ScrollContainer" needsEntityContext="true" offlineCapable="true"
        pluginWidget="true"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Scroll container</name>
    <description>Displays a scrollable container</description>
    <studioProCategory>Layout</studioProCategory>
    <studioCategory>Layout</studioCategory>
    <icon>iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAHkSURBVHgBvVVNTsJAFH4D3SguvIG9gXgDXBirK5tAws5yAvEE1BMAJ6CuISndCO64gXgDjsACpRrK+D2YRiDaTkn0S15npm/e+zrT90P0xxD8qFar5mKxcOKXmLf6/f6UMqJcLtcxHCsfHnxMDLVwhBANtY8de2rURqVSKWJoxmvDWLl2jc1N3W5XJDmx+vMOj8Obg9quDrZjUjcCMhm/z5EmrOCDT+iwWMG8oWunRXAZhHckl66Usi0gJMm1gs9bHdtUgmt/VoTTFqajZ/uwPoDwnGTkXUCXZm8kKS1/bi4F+Wo5sfw3lycgfJVClPIi72PP+dA+mOxFAJQgppo7JNYHlt96U+3xaB8CfJmXZKwD7ShiIPwcliw22gTKMedBJwtJphPsA20CZKqHgTO4puZaSIuin0gyISUPZgjNfCdpD2K2pqItOwFRfoTHBGKuygTJdYUVuRNa1yXWjZI8JP4DztCcjOyVTyFOh3bBZSGVfKxLymKNExA92UdjlIMaCnHnKnhv8jfhNCWUi3vWpdlr/WS+Y65LkrgpoVDI5cPALrR0bLXDFCQuHD+yqGvSwtYJNjrRmepQOyQFh34B9/Uoil5I9eQtAvRPD305fjft9Xqpd7uLMAyn8NOO19z06T/wBefnxUyC/zU5AAAAAElFTkSuQmCC</icon>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Width">
                <property key="widthType" type="enumeration" defaultValue="full">
                    <caption>Width</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="full">Full width</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="widthPixels" type="decimal" defaultValue="500">
                    <caption>Pixels</caption>
                    <description>The amount of pixels to be considered as width.</description>
                </property>
                <property key="widthPercentage" type="decimal" defaultValue="100">
                    <caption>Percentage</caption>
                    <description>The amount of percentage to be considered as width.</description>
                </property>
                <property key="content" type="widgets">
                    <caption>Content</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
