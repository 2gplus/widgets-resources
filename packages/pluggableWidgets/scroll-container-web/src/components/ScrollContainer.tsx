import { createElement, CSSProperties, PropsWithChildren, ReactElement, useCallback, useEffect, useState } from "react";
import classNames from "classnames";

interface ScrollContainerProps {
    className: string;
    width?: number;
    percentage?: number;
    styles?: CSSProperties;
}

export function ScrollContainer(props: PropsWithChildren<ScrollContainerProps>): ReactElement {
    const [divContainer, setDivContainer] = useState<HTMLDivElement | null>(null);
    const [renderContents, showRenderContents] = useState(false);
    const [height, setHeight] = useState<number | undefined>(undefined);

    const handleResize = useCallback(() => {
        if (divContainer?.parentElement?.className.match(/mx-placeholder|region-content/)) {
            const topBar = document.querySelector(".topbar-content");
            setHeight(window.innerHeight - (topBar?.clientHeight ?? 0));
        } else {
            // Impossible to define height of the element
            showRenderContents(true);
        }
    }, [divContainer]);

    useEffect(() => {
        if (divContainer?.clientHeight === 0) {
            handleResize();
        }
    }, [divContainer, handleResize]);

    useEffect(() => {
        window.addEventListener("resize", handleResize);
        return () => {
            window.removeEventListener("resize", handleResize);
        };
    }, [handleResize]);

    return (
        <div
            ref={ref => {
                if (ref && !divContainer) {
                    setDivContainer(ref);
                }
            }}
            className={classNames("widget-scroll-container", props.className)}
            style={{
                ...(props.width
                    ? { width: `${props.width}px` }
                    : props.percentage
                    ? { width: `${props.percentage}%` }
                    : undefined),
                ...(height !== undefined ? { height } : undefined),
                ...props.styles
            }}
        >
            <div
                className="widget-scroll-container-wrapper"
                style={renderContents ? { display: "contents" } : undefined}
            >
                {props.children}
            </div>
        </div>
    );
}
