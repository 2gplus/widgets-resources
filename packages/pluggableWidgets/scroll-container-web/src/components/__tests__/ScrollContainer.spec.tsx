import { createElement } from "react";
import { shallow } from "enzyme";
import { ScrollContainer } from "../ScrollContainer";

describe("Scroll Container", () => {
    it("renders correctly the structure", () => {
        const scrollContainer = shallow(<ScrollContainer className="scroll-container-test" />);
        expect(scrollContainer).toMatchSnapshot();
    });
});
