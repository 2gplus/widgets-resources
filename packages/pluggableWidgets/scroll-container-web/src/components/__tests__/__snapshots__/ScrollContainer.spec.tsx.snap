// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scroll Container renders correctly the structure 1`] = `
<div
  className="widget-scroll-container scroll-container-test"
  style={Object {}}
>
  <div
    className="widget-scroll-container-wrapper"
  />
</div>
`;
